data/transactions/logic/resources.go

```diff
@@ -0,0 +1,359 @@
+	}
+	err = cx.requireHolding(tx.AssetSender, tx.XferAsset)
+	if err != nil {
+		return fmt.Errorf("axfer AssetSender: %w", err)
+	}
+	err = cx.requireHolding(tx.AssetCloseTo, tx.XferAsset)
+	if err != nil {
+		return fmt.Errorf("axfer AssetCloseTo: %w", err)
+	}
+	return nil
+}
+
+func (r *resources) fillAssetFreeze(hdr *transactions.Header, tx *transactions.AssetFreezeTxnFields) {
+	r.sharedAccounts[hdr.Sender] = struct{}{}
+	id := tx.FreezeAsset
+	r.sharedAsas[id] = struct{}{}
+	r.shareAccountAndHolding(tx.FreezeAccount, id)
+}
+
+func (cx *EvalContext) allowsAssetFreeze(hdr *transactions.Header, tx *transactions.AssetFreezeTxnFields) error {
+	err := cx.requireHolding(tx.FreezeAccount, tx.FreezeAsset)
+	if err != nil {
+		return fmt.Errorf("afrz FreezeAccount: %w", err)
+	}
+	return nil
+}
+
+func (r *resources) fillApplicationCall(ep *EvalParams, hdr *transactions.Header, tx *transactions.ApplicationCallTxnFields) {
+	txAccounts := make([]basics.Address, 0, 2+len(tx.Accounts)+len(tx.ForeignApps))
+	txAccounts = append(txAccounts, hdr.Sender)
+	txAccounts = append(txAccounts, tx.Accounts...)
+	for _, id := range tx.ForeignAssets {
+		r.sharedAsas[id] = struct{}{}
+	}
+	// Make the app account associated with app calls available. We
+	// don't have to add code to make the accounts of freshly created
+	// apps available, because that is already handled by looking at
+	// `createdApps`.
+	if id := tx.ApplicationID; id != 0 {
+		txAccounts = append(txAccounts, ep.getApplicationAddress(id))
+		r.sharedApps[id] = struct{}{}
+	}
+	for _, id := range tx.ForeignApps {
+		txAccounts = append(txAccounts, ep.getApplicationAddress(id))
+		r.sharedApps[id] = struct{}{}
+	}
+	for _, address := range txAccounts {
+		r.sharedAccounts[address] = struct{}{}
+
+		for _, id := range tx.ForeignAssets {
+			r.shareHolding(address, id)
+		}
+		// Similar to note about app accounts, availableLocals allows
+		// all createdApps holdings, so we don't care if id == 0 here.
+		if id := tx.ApplicationID; id != 0 {
+			r.shareLocal(address, id)
+		}
+		for _, id := range tx.ForeignApps {
+			r.shareLocal(address, id)
+		}
+	}
+
+	for _, br := range tx.Boxes {
+		var app basics.AppIndex
+		if br.Index == 0 {
+			// "current app": Ignore if this is a create, else use ApplicationID
+			if tx.ApplicationID == 0 {
+				// When the create actually happens, and we learn the appID, we'll add it.
+				continue
+			}
+			app = tx.ApplicationID
+		} else {
+			// Bounds check will already have been done by
+			// WellFormed. For testing purposes, it's better to panic
+			// now than after returning a nil.
+			app = tx.ForeignApps[br.Index-1] // shift for the 0=this convention
+		}
+		r.boxes[boxRef{app, string(br.Name)}] = false
+	}
+}
+
+func (cx *EvalContext) allowsApplicationCall(hdr *transactions.Header, tx *transactions.ApplicationCallTxnFields, callerVer, calleeVer uint64) error {
+	// If an old (pre resource sharing) app is being called from an app that has
```

Comment:
these are cases where the caller is the sharing version or better but the application called is not the sharing version.